core/src/main/java/io/parsingdata/metal/data/ParseState.java

```diff
@@ -36,43 +37,55 @@
     public final ParseGraph order;
     public final BigInteger offset;
     public final Source source;
+    public final ImmutableList<BigInteger> iterations;
 
-    public ParseState(final ParseGraph order, final Source source, final BigInteger offset) {
+    public ParseState(final ParseGraph order, final Source source, final BigInteger offset, final ImmutableList<BigInteger> iterations) {
         this.order = checkNotNull(order, "order");
         this.source = checkNotNull(source, "source");
         this.offset = checkNotNegative(offset, "offset");
+        this.iterations = checkNotNull(iterations, "iteration");
     }
 
     public static ParseState createFromByteStream(final ByteStream input, final BigInteger offset) {
-        return new ParseState(ParseGraph.EMPTY, new ByteStreamSource(input), offset);
+        return new ParseState(ParseGraph.EMPTY, new ByteStreamSource(input), offset, new ImmutableList<>());
     }
 
     public static ParseState createFromByteStream(final ByteStream input) {
         return createFromByteStream(input, ZERO);
     }
 
     public ParseState addBranch(final Token token) {
-        return new ParseState(order.addBranch(token), source, offset);
+        if (token.isIterable()) {
```

Comment:
This could be a oneliner... Does @jvdb approve of this heresy? 🤣

Reply:
Yes, it _could_, but _should_ it? ;) The code now clearly shows the difference between the two states, which will not be as clear if it is a oneliner. @jvdb will approve this, I'm sure! :D

Reply:
To be entirely honest, seeing multiple identical `addBranch()`-invocations does hurt a little... But I'm not religious about this :)

Reply:
I would prefer that yes :+1: